src/file_io.jl

```diff
@@ -107,3 +125,26 @@ end
 # FileIO Interface
 fileio_save(f, args...; kwargs...) = save(f.filename, args...; kwargs...)
 fileio_load(f, args...) = load(f.filename, args...)
+
+# Objects were initialially saved inline with the BSON document prior to v4
+# To support larger objects/files we started writing objects to the end of file and only save
+# names and sizes to the objects dict in the bson doc.
+# Since our ideal structure is one dict with object bytes we don't upgrade, but rather
+# extract both formats.
+function _extract_objects(io::IO, d)
+    if VersionNumber(d["metadata"]["version"]) < v"4"
+        return Dict{Symbol, Vector{UInt8}}(Symbol(k) => v for (k, v) in d["objects"])
+    else
+        objects = Dict{Symbol, Vector{UInt8}}()
+
+        for (k, nb) in zip(d["object-names"], d["object-nbytes"])
+            bytes = read(io, nb)
+            if length(bytes) < nb
+                throw(ArgumentError("Failed to read all $nb bytes for object $k"))
```

Comment:
This is not an `ArgumentError`.
it is an `EOFError` though that doesn't have ability to attach a message.
OTOH I am not sure that that message is particularly useful, it might be or it might not be.
Do we exect this to occur because the file has been reandomly truncated (if so `EOFError` is fine)
or because some particular data will break the stream some how during righting (in which case we want to report the error)
We could strike a middle ground by logging something and using a `EOFError`.

(In general this kind of thing generally makes me thing julia (or any other programming language I've used) hasn't yet found the best way to do exceptions)

Reply:
Yeah, unfortunately, this is trying to do both. I suppose the most likely error for writing issues is that the object ordering could get messed up, which this wouldn't catch anyways. I'll just change it to a `EOFError`.